In [20]:
import pandas as pd
import numpy as np
import plotly.express as px
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime
import pycountry

In [21]:
df = pd.read_csv("/Users/fersaol/Library/CloudStorage/OneDrive-Personal/Documentos/MIS_COSAS_DE_THE_BRIDGE/EDA/proyecto_co2/src/data/copia_seguridad_1.csv")
df.head()

Unnamed: 0  index        Year Country           GDP    Population  \
0           0      0  1980-01-01   World  27770.910281  4.298127e+06   
1           1      1  1980-01-01   World  27770.910281  4.298127e+06   
2           2      2  1980-01-01   World  27770.910281  4.298127e+06   
3           3      3  1980-01-01   World  27770.910281  4.298127e+06   
4           4      4  1980-01-01   World  27770.910281  4.298127e+06   

   Energy_production  Energy_consumption                Energy_type  \
0         296.337228          292.899790           all_energy_types   
1          20.775178           20.702344         renewables_n_other   
2           7.575700            7.575700                    nuclear   
3         133.111109          132.064019  petroleum_n_other_liquids   
4          54.761046           53.865223                natural_gas   

   CO2_emission  ...   balance  eficiency  energy_dependecy  use_intensity_pc  \
0   4946.627130  ...  3.437438  16.692561          0.010547          0.000068   
1      0.000000  ...  0.072834   0.000000          0.000745          0.000005   
2      0.000000  ...  0.000000   0.000000          0.000273          0.000002   
3   2455.243565  ...  1.047089  18.445069          0.004755          0.000031   
4   1081.593377  ...  0.895822  19.751145          0.001940          0.000013   

     co2_pc  CODE  continent geometry latitude longitude  
0  0.001151  None        NaN      NaN      NaN       NaN  
1  0.000000  None        NaN      NaN      NaN       NaN  
2  0.000000  None        NaN      NaN      NaN       NaN  
3  0.000571  None        NaN      NaN      NaN       NaN  
4  0.000252  None        NaN      NaN      NaN       NaN  

[5 rows x 22 columns]

Eliminamos las columnas no necesarias que se han creado con los reset_index:

In [22]:
columnas = df.columns
columnas

Index(['Unnamed: 0', 'index', 'Year', 'Country', 'GDP', 'Population',
       'Energy_production', 'Energy_consumption', 'Energy_type',
       'CO2_emission', 'per_capita_production', 'Energy_intensity_by_GDP',
       'balance', 'eficiency', 'energy_dependecy', 'use_intensity_pc',
       'co2_pc', 'CODE', 'continent', 'geometry', 'latitude', 'longitude'],
      dtype='object')

In [23]:
df = df.drop(columns=columnas[:2])
df.head()

Year Country           GDP    Population  Energy_production  \
0  1980-01-01   World  27770.910281  4.298127e+06         296.337228   
1  1980-01-01   World  27770.910281  4.298127e+06          20.775178   
2  1980-01-01   World  27770.910281  4.298127e+06           7.575700   
3  1980-01-01   World  27770.910281  4.298127e+06         133.111109   
4  1980-01-01   World  27770.910281  4.298127e+06          54.761046   

   Energy_consumption                Energy_type  CO2_emission  \
0          292.899790           all_energy_types   4946.627130   
1           20.702344         renewables_n_other      0.000000   
2            7.575700                    nuclear      0.000000   
3          132.064019  petroleum_n_other_liquids   2455.243565   
4           53.865223                natural_gas   1081.593377   

   per_capita_production  Energy_intensity_by_GDP   balance  eficiency  \
0              68.145921                   10.547  3.437438  16.692561   
1              68.145921                   10.547  0.072834   0.000000   
2              68.145921                   10.547  0.000000   0.000000   
3              68.145921                   10.547  1.047089  18.445069   
4              68.145921                   10.547  0.895822  19.751145   

   energy_dependecy  use_intensity_pc    co2_pc  CODE continent geometry  \
0          0.010547          0.000068  0.001151  None       NaN      NaN   
1          0.000745          0.000005  0.000000  None       NaN      NaN   
2          0.000273          0.000002  0.000000  None       NaN      NaN   
3          0.004755          0.000031  0.000571  None       NaN      NaN   
4          0.001940          0.000013  0.000252  None       NaN      NaN   

   latitude  longitude  
0       NaN        NaN  
1       NaN        NaN  
2       NaN        NaN  
3       NaN        NaN  
4       NaN        NaN

In [24]:
df.index

RangeIndex(start=0, stop=44320, step=1)

Vemos que del dataframe original que tenía 44278 ahora tenemos 44320 y es porque se nos han creado filas llenas de nan al final las cuales pasamos a eliminar

In [25]:
df.tail(50)

Year                   Country      GDP  Population  \
44270  2018-01-01               South Sudan  17.1853     10975.9   
44271  2018-01-01               South Sudan  17.1853     10975.9   
44272  2018-01-01               South Sudan  17.1853     10975.9   
44273  2019-01-01               South Sudan  17.3349     11062.1   
44274  2019-01-01               South Sudan  17.3349     11062.1   
44275  2019-01-01               South Sudan  17.3349     11062.1   
44276  2019-01-01               South Sudan  17.3349     11062.1   
44277  2019-01-01               South Sudan  17.3349     11062.1   
44278         NaN                       NaN      NaN         NaN   
44279         NaN                       NaN      NaN         NaN   
44280         NaN                       NaN      NaN         NaN   
44281         NaN                       NaN      NaN         NaN   
44282         NaN                       NaN      NaN         NaN   
44283         NaN                       NaN      NaN         NaN   
44284         NaN                       NaN      NaN         NaN   
44285         NaN                       NaN      NaN         NaN   
44286         NaN                       NaN      NaN         NaN   
44287         NaN                       NaN      NaN         NaN   
44288         NaN                       NaN      NaN         NaN   
44289         NaN                       NaN      NaN         NaN   
44290         NaN                       NaN      NaN         NaN   
44291         NaN                       NaN      NaN         NaN   
44292         NaN                       NaN      NaN         NaN   
44293         NaN                       NaN      NaN         NaN   
44294         NaN                       NaN      NaN         NaN   
44295         NaN                       NaN      NaN         NaN   
44296         NaN                       NaN      NaN         NaN   
44297         NaN                       NaN      NaN         NaN   
44298         NaN                       NaN      NaN         NaN   
44299         NaN                       NaN      NaN         NaN   
44300         NaN                       NaN      NaN         NaN   
44301         NaN                       NaN      NaN         NaN   
44302         NaN                       NaN      NaN         NaN   
44303         NaN                       NaN      NaN         NaN   
44304         NaN                       NaN      NaN         NaN   
44305         NaN                       NaN      NaN         NaN   
44306         NaN                   Bahamas      NaN         NaN   
44307         NaN          Bosnia and Herz.      NaN         NaN   
44308         NaN      Central African Rep.      NaN         NaN   
44309         NaN                     Congo      NaN         NaN   
44310         NaN                   Czechia      NaN         NaN   
44311         NaN                    Côte d      NaN         NaN   
44312         NaN           Dem. Rep. Congo      NaN         NaN   
44313         NaN            Dominican Rep.      NaN         NaN   
44314         NaN                Eq. Guinea      NaN         NaN   
44315         NaN                    Gambia      NaN         NaN   
44316         NaN                 Macedonia      NaN         NaN   
44317         NaN                  S. Sudan      NaN         NaN   
44318         NaN  United States of America      NaN         NaN   
44319         NaN                 W. Sahara      NaN         NaN   

       Energy_production  Energy_consumption                Energy_type  \
44270           0.000000            0.000000                natural_gas   
44271           0.289098            0.023736  petroleum_n_other_liquids   
44272           0.000036            0.000036         renewables_n_other   
44273           0.291579            0.026594           all_energy_types   
44274           0.000000            0.000000                natural_gas   
44275           0.000000            0.000000                       coal   
44276           0.0

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44320 entries, 0 to 44319
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     44278 non-null  object 
 1   Country                  44292 non-null  object 
 2   GDP                      44278 non-null  float64
 3   Population               44278 non-null  float64
 4   Energy_production        44278 non-null  float64
 5   Energy_consumption       44278 non-null  float64
 6   Energy_type              44278 non-null  object 
 7   CO2_emission             44278 non-null  float64
 8   per_capita_production    44278 non-null  float64
 9   Energy_intensity_by_GDP  44278 non-null  float64
 10  balance                  44278 non-null  float64
 11  eficiency                44278 non-null  float64
 12  energy_dependecy         44278 non-null  float64
 13  use_intensity_pc         44278 non-null  float64
 14  co2_pc                

In [27]:
df = df.iloc[:44278,:]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44278 entries, 0 to 44277
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     44278 non-null  object 
 1   Country                  44278 non-null  object 
 2   GDP                      44278 non-null  float64
 3   Population               44278 non-null  float64
 4   Energy_production        44278 non-null  float64
 5   Energy_consumption       44278 non-null  float64
 6   Energy_type              44278 non-null  object 
 7   CO2_emission             44278 non-null  float64
 8   per_capita_production    44278 non-null  float64
 9   Energy_intensity_by_GDP  44278 non-null  float64
 10  balance                  44278 non-null  float64
 11  eficiency                44278 non-null  float64
 12  energy_dependecy         44278 non-null  float64
 13  use_intensity_pc         44278 non-null  float64
 14  co2_pc                

In [28]:
df.tail(2)

Year      Country      GDP  Population  Energy_production  \
44276  2019-01-01  South Sudan  17.3349     11062.1           0.000036   
44277  2019-01-01  South Sudan  17.3349     11062.1           0.291543   

       Energy_consumption                Energy_type  CO2_emission  \
44276            0.000036         renewables_n_other      0.000000   
44277            0.026558  petroleum_n_other_liquids      1.675889   

       per_capita_production  Energy_intensity_by_GDP   balance  eficiency  \
44276               2.404033                 1.534111  0.000000   0.000000   
44277               2.404033                 1.534111  0.264985   5.748336   

       energy_dependecy  use_intensity_pc    co2_pc CODE continent  \
44276          0.000002      3.219642e-09  0.000000  SSD    Africa   
44277          0.001532      2.400814e-06  0.000151  SSD    Africa   

                                                geometry  latitude  longitude  
44276  POLYGON ((30.833852421715427 3.509171604222462...       NaN        NaN  
44277  POLYGON ((30.833852421715427 3.509171604222462...       NaN        NaN

Sustituimos los datos con infinito en la columna de eficiency por 0:

In [29]:
len(df[df.eficiency == np.inf])

8561

In [30]:
df["eficiency"] = df["eficiency"].replace(np.inf,0)
df[df.eficiency == np.inf]

Empty DataFrame
Columns: [Year, Country, GDP, Population, Energy_production, Energy_consumption, Energy_type, CO2_emission, per_capita_production, Energy_intensity_by_GDP, balance, eficiency, energy_dependecy, use_intensity_pc, co2_pc, CODE, continent, geometry, latitude, longitude]
Index: []

### Qué países son los que no tienen datos de localización?

In [31]:
df.Country[(df.latitude.isnull()) | (df.longitude.isnull())].unique()

array(['World', 'North Korea', 'Netherlands Antilles', 'Reunion',
       'Saint Helena', 'Macau', 'Micronesia', 'U.S. Pacific Islands',
       'U.S. Virgin Islands', 'U.S. Territories', 'The Bahamas',
       'Wake Island', 'Saint Vincent/Grenadines', 'Burma',
       'Côte d’Ivoire', 'Congo-Kinshasa', 'Congo-Brazzaville',
       'British Virgin Islands', 'Hawaiian Trade Zone',
       'Falkland Islands', 'Former Czechoslovakia', 'Former Yugoslavia',
       'Gambia, The', 'Former Serbia and Montenegro', 'Czech Republic',
       'Palestinian Territories', 'Kosovo', 'Northern Mariana Islands',
       'Norway', 'Niue', 'New Caledonia', 'Puerto Rico', 'Malta',
       'Maldives', 'Lesotho', 'Saint Kitts and Nevis', 'Martinique',
       'Montserrat', 'Nauru', 'Mauritius', 'Tuvalu',
       'Turks and Caicos Islands', 'Tonga', 'Western Sahara', 'Vanuatu',
       'United States', 'Seychelles', 'Singapore', 'Solomon Islands',
       'Saint Pierre and Miquelon', 'Saint Lucia', 'Samoa',
       'Sao T

In [32]:
df[df.Country == "United States"]

Year        Country       GDP  Population  Energy_production  \
18942  1980-01-01  United States   7080.75    227119.0          67.146595   
18943  1980-01-01  United States   7080.75    227119.0          18.597725   
18944  1980-01-01  United States   7080.75    227119.0          19.907600   
18945  1980-01-01  United States   7080.75    227119.0           2.739169   
18946  1980-01-01  United States   7080.75    227119.0          20.473758   
...           ...            ...       ...         ...                ...   
19177  2019-01-01  United States  19925.43    330043.8          35.187183   
19178  2019-01-01  United States  19925.43    330043.8          14.255763   
19179  2019-01-01  United States  19925.43    330043.8          11.595138   
19180  2019-01-01  United States  19925.43    330043.8           8.451852   
19181  2019-01-01  United States  19925.43    330043.8          31.910982   

       Energy_consumption                Energy_type  CO2_emission  \
18942           78.021113           all_energy_types   4946.627130   
18943           15.422809                       coal   1409.790188   
18944           20.235459                natural_gas   1081.593377   
18945            2.739169                    nuclear      0.000000   
18946           34.158965  petroleum_n_other_liquids   2455.243565   
...                   ...                        ...           ...   
19177           32.252007                natural_gas   1626.410978   
19178           11.315595                       coal   1262.555409   
19179           11.436136         renewables_n_other      0.000000   
19180            8.451852                    nuclear      0.000000   
19181           38.329386  petroleum_n_other_liquids   2388.226367   

       per_capita_production  Energy_intensity_by_GDP    balance   eficiency  \
18942             343.525258                11.018764 -10.874518   73.669069   
18943             343.525258                11.018764   3.174916   75.804441   
18944             343.525258                11.018764  -0.327859   54.330676   
18945             343.525258                11.018764   0.000000    0.000000   
18946             343.525258                11.018764 -13.685207  119.921490   
...                      ...                      ...        ...         ...   
19177             304.305271                 5.040497   2.935176   46.221687   
19178             304.305271                 5.040497   2.940168   88.564562   
19179             304.305271                 5.040497   0.159002    0.000000   
19180             304.305271                 5.040497   0.000000    0.000000   
19181             304.305271                 5.040497  -6.418404   74.840266   

       energy_dependecy  use_intensity_pc    co2_pc CODE      continent  \
18942          0.011019          0.000344  0.021780  USA  North America   
18943          0.002178          0.000068  0.006207  USA  North America   
18944          0.002858          0.000089  0.004762  USA  North America   
18945          0.000387          0.000012  0.000000  USA  North America   
18946          0.004824          0.000150  0.010810  USA  North America   
...                 ...               ...       ...  ...            ...   
19177          0.001619          0.000098  0.004928  USA  North America   
19178          0.000568          0.000034  0.003825  USA  North America   
19179          0.000574          0.000035  0.000000  USA  North America   
19180          0.000424          0.000026  0.000000  USA  North America   
19181          0.001924          0.000116  0.007236  USA  North America   

                                                geometry  latitude  longitude  
18942  MULTIPOLYGON (((-122.84000000000003 49.0000000...       NaN        NaN  
18943  MULTIPOLYGON (((-122.84000000000003 49.0000000...       NaN        NaN  
18944  MULTIPOLYGON (((-122.84000000000003 49.0000000...       NaN        NaN  
18945  MULTIPOLYGON (((-122.84000000000003 49.0000000...       Na

El caso es que en el dataframe con el que hicimos el merge si que teníamos los datos de localización de Estados Unidos, pero por alguna razón se han destruido

In [33]:
git_locations = pd.read_csv("/Users/fersaol/Library/CloudStorage/OneDrive-Personal/Documentos/MIS_COSAS_DE_THE_BRIDGE/EDA/proyecto_co2/src/data/location_countries.csv")
git_locations.head(10)

ISO 3166 Country Code               Country  Latitude  Longitude
0                    AD               Andorra     42.50       1.50
1                    AE  United Arab Emirates     24.00      54.00
2                    AF           Afghanistan     33.00      65.00
3                    AG   Antigua and Barbuda     17.05     -61.80
4                    AI              Anguilla     18.25     -63.17
5                    AL               Albania     41.00      20.00
6                    AM               Armenia     40.00      45.00
7                    AN  Netherlands Antilles     12.25     -68.75
8                    AO                Angola    -12.50      18.50
9                    AP   Asia/Pacific Region     35.00     105.00

Vemos que el dataframe anterior tiene los datos de longitud y latitud promediados y por eso buscamos uno alternativo encontrandolo en una url de google:

In [34]:
location = pd.read_html("https://developers.google.com/public-data/docs/canonical/countries_csv")
type(location)

list

Vemos que es de tipo lista porque al intentar abrirlo como dataframe nos decía que no era de dos dimensiones, por lo que lo pasamos a un array de numpy

In [35]:
location = np.array(location)

vemos que no es de dos dimensiones sino que tiene cuatro, por lo que le hacemos un reshape

In [36]:
location.shape

(1, 245, 4)

In [37]:
location = location.reshape(245,4)

Convertimos el array de numpy en dataframe y renombramos las columnas:

In [38]:
df_locations = pd.DataFrame(location)
df_locations = df_locations.rename(columns={0:"CODE",1:"latitude",2:"longitude",3:"Country"})
df_locations.head()

CODE   latitude  longitude               Country
0   AD  42.546245   1.601554               Andorra
1   AE  23.424076  53.847818  United Arab Emirates
2   AF   33.93911  67.709953           Afghanistan
3   AG  17.060816 -61.796428   Antigua and Barbuda
4   AI  18.220554 -63.068615              Anguilla

Eliminamos las columnas antiguas:

In [40]:
df = df.drop(columns=["longitude","latitude"])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44278 entries, 0 to 44277
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     44278 non-null  object 
 1   Country                  44278 non-null  object 
 2   GDP                      44278 non-null  float64
 3   Population               44278 non-null  float64
 4   Energy_production        44278 non-null  float64
 5   Energy_consumption       44278 non-null  float64
 6   Energy_type              44278 non-null  object 
 7   CO2_emission             44278 non-null  float64
 8   per_capita_production    44278 non-null  float64
 9   Energy_intensity_by_GDP  44278 non-null  float64
 10  balance                  44278 non-null  float64
 11  eficiency                44278 non-null  float64
 12  energy_dependecy         44278 non-null  float64
 13  use_intensity_pc         44278 non-null  float64
 14  co2_pc                

juntamos ambos dataframes

In [41]:
df = pd.merge(df,df_locations,on="Country",how="outer")
df.head()

Year Country           GDP    Population  Energy_production  \
0  1980-01-01   World  27770.910281  4.298127e+06         296.337228   
1  1980-01-01   World  27770.910281  4.298127e+06          20.775178   
2  1980-01-01   World  27770.910281  4.298127e+06           7.575700   
3  1980-01-01   World  27770.910281  4.298127e+06         133.111109   
4  1980-01-01   World  27770.910281  4.298127e+06          54.761046   

   Energy_consumption                Energy_type  CO2_emission  \
0          292.899790           all_energy_types   4946.627130   
1           20.702344         renewables_n_other      0.000000   
2            7.575700                    nuclear      0.000000   
3          132.064019  petroleum_n_other_liquids   2455.243565   
4           53.865223                natural_gas   1081.593377   

   per_capita_production  Energy_intensity_by_GDP  ...  eficiency  \
0              68.145921                   10.547  ...  16.692561   
1              68.145921                   10.547  ...   0.000000   
2              68.145921                   10.547  ...   0.000000   
3              68.145921                   10.547  ...  18.445069   
4              68.145921                   10.547  ...  19.751145   

   energy_dependecy  use_intensity_pc    co2_pc  CODE_x continent geometry  \
0          0.010547          0.000068  0.001151    None       NaN      NaN   
1          0.000745          0.000005  0.000000    None       NaN      NaN   
2          0.000273          0.000002  0.000000    None       NaN      NaN   
3          0.004755          0.000031  0.000571    None       NaN      NaN   
4          0.001940          0.000013  0.000252    None       NaN      NaN   

  CODE_y latitude longitude  
0    NaN      NaN       NaN  
1    NaN      NaN       NaN  
2    NaN      NaN       NaN  
3    NaN      NaN       NaN  
4    NaN      NaN       NaN  

[5 rows x 21 columns]

Comprobamos como ha quedado nuestro dataframe

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44317 entries, 0 to 44316
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     44278 non-null  object 
 1   Country                  44317 non-null  object 
 2   GDP                      44278 non-null  float64
 3   Population               44278 non-null  float64
 4   Energy_production        44278 non-null  float64
 5   Energy_consumption       44278 non-null  float64
 6   Energy_type              44278 non-null  object 
 7   CO2_emission             44278 non-null  float64
 8   per_capita_production    44278 non-null  float64
 9   Energy_intensity_by_GDP  44278 non-null  float64
 10  balance                  44278 non-null  float64
 11  eficiency                44278 non-null  float64
 12  energy_dependecy         44278 non-null  float64
 13  use_intensity_pc         44278 non-null  float64
 14  co2_pc                

Vemos que los códigos de países nuevos tienen menos datos por lo que lo eliminamos y nos quedamos con el antiguo:

In [43]:
df = df.drop(columns="CODE_y")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44317 entries, 0 to 44316
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     44278 non-null  object 
 1   Country                  44317 non-null  object 
 2   GDP                      44278 non-null  float64
 3   Population               44278 non-null  float64
 4   Energy_production        44278 non-null  float64
 5   Energy_consumption       44278 non-null  float64
 6   Energy_type              44278 non-null  object 
 7   CO2_emission             44278 non-null  float64
 8   per_capita_production    44278 non-null  float64
 9   Energy_intensity_by_GDP  44278 non-null  float64
 10  balance                  44278 non-null  float64
 11  eficiency                44278 non-null  float64
 12  energy_dependecy         44278 non-null  float64
 13  use_intensity_pc         44278 non-null  float64
 14  co2_pc                

### Los nuevos países sin datos de localización son

In [44]:
df.Country[(df.latitude.isnull()) | (df.longitude.isnull())].unique()

array(['World', 'Reunion', 'U.S. Pacific Islands', 'U.S. Territories',
       'The Bahamas', 'Wake Island', 'Saint Vincent/Grenadines', 'Burma',
       'Côte d’Ivoire', 'Congo-Kinshasa', 'Congo-Brazzaville',
       'Hawaiian Trade Zone', 'Falkland Islands', 'Former Czechoslovakia',
       'Former Yugoslavia', 'Gambia, The', 'Former Serbia and Montenegro',
       'Sao Tome and Principe', 'Cabo Verde', 'Eswatini',
       'North Macedonia', 'South Sudan', 'U.S. Minor Outlying Islands'],
      dtype=object)

In [46]:
len(df.Country.unique())

267

### Realizamos una copia de seguridad de los datos

In [48]:
df.to_csv("copia_de_seguridad_2.csv")